### PR TITLE
Logging with composite pattern

### PR DESCRIPTION
**Pull request by RabeaGleissner**
@felipesere @jsuchy

This PR implements
- the logger writing the logs to a file
- the logger printing the logs to the console

I used the composite pattern as recommended by Felipe.

### CODE REVIEW

File: src/main/java/de/rabea/logger/ConsoleLogger.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    @Override
+    public void log(String message) {
+        console.write(message);
+    }
+
+    @Override
+    public String getLogs() {
+        return "";
```
Comment (RabeaGleissner) [inline src/main/java/de/rabea/logger/ConsoleLogger.java:20]: @felipesere I had to implement this method as part of the Interface to return the logs - but it only works with the FileLogger. So I'm just returning an empty string here but I feel that that's not ideal. Do you have any other suggestions how to resolve this?

File: src/test/java/de/rabea/controller/LogsControllerTest.java [hunk 2]
```diff
@@ -14,22 +20,34 @@
...
+        File file = folder.newFile("log.txt");
+        pathToEmptyFile = file.getAbsolutePath();
+    }
+
     @Test
     public void returnsResponseWithLogsIfAuthorised() {
-        Logger logger = new Logger();
-        logger.log(new RequestLine("GET /somewhere HTTP/1.1"));
+        MultiLogger logger = new MultiLogger();
+        logger.add(new FileLogger(pathToEmptyFile));
```
Comment (RabeaGleissner) [inline src/test/java/de/rabea/controller/LogsControllerTest.java]: @felipesere
You have to know that you need to add different loggers to the `MultiLogger`, both in the tests and also in Main. Is that just how the pattern works or are there any ways to make it more obvious that other loggers need to be added?

File: src/main/java/de/rabea/Main.java [hunk 3]
```diff
@@ -16,7 +24,10 @@ public static void main(String[] args) throws IOException {
 
         ServerSocket serverSocket = new ServerSocket(port);
         Router router = new Router();
-        Logger logger = new Logger();
+
+        MultiLogger logger = new MultiLogger(new ArrayList<>(Arrays.asList(new FileLogger(logFilePath),
```
Comment (felipesere) [inline src/main/java/de/rabea/Main.java]: This shouldn't need to be a `MultiLogger` on the left-hand side of the assignment... nobody should know what kind of logger they are talking to...

File: src/main/java/de/rabea/controller/LogsController.java [hunk 4]
```diff
@@ -12,17 +12,17 @@
 
 public class LogsController extends Controller {
 
-    private final Logger logger;
+    private final MultiLogger logger;
 
-    public LogsController(Logger logger) {
+    public LogsController(MultiLogger logger) {
```
Comment (felipesere) [inline src/main/java/de/rabea/controller/LogsController.java]: No need use a specific kind of `Logger` here either...

File: src/main/java/de/rabea/logger/FileLogger.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        return new FileWriter(filePath);
+    }
+
+    private String formatted(String message) {
+        return message + ", ";
+    }
+
+    public class FileWriterException extends RuntimeException {
+        public FileWriterException() {
+            super("Sorry, the FileWriter could not be created");
```
Comment (felipesere) [inline src/main/java/de/rabea/logger/FileLogger.java]: Is this exception really helpful? You are dropping the actual stacktrace.
Just pass the original exception as a parameter...

File: src/main/java/de/rabea/Main.java (hunk 3 above)
Comment (RabeaGleissner) [inline src/main/java/de/rabea/Main.java]: OK, I've changed it to use the Interface here.

File: src/main/java/de/rabea/controller/LogsController.java (hunk 4 above)
Comment (RabeaGleissner) [inline src/main/java/de/rabea/controller/LogsController.java]: Updated.

File: src/main/java/de/rabea/logger/FileLogger.java (hunk 5 above)
Comment (RabeaGleissner) [inline src/main/java/de/rabea/logger/FileLogger.java]: Updated to pass in the `e.getMessage()`

**Pull request merged**